src/main/Makefile.am

```diff
@@ -28,10 +28,11 @@ pcvc4_SOURCES = \
 	driver_unified.cpp
 pcvc4_LDADD = \
 	libmain.a \
-	@builddir@/../parser/libcvc4parser.la \
-	@builddir@/../libcvc4.la \
 	$(READLINE_LIBS) \
-	@builddir@/../lib/libreplacements.la
+	@builddir@/../lib/libreplacements.la \
+	@builddir@/../libcvc4.la \
+	@builddir@/../parser/libcvc4parser.la \
+	@builddir@/../util/libutil.la
```

Comment:
After reading through the links, it seems like `cvc4_private.h` and `cvc4_private_library.h` basically define when certain headers can be used. E.g. headers that include `cvc4_private.h` can only be used when building the CVC4 library whereas `cvc4_private_library.h` indicates that the header can also be used in the driver and parsers. StatisticsRegistry and friends are used in the driver, so we can't make them private. We also don't want to make them public to anyone else linking against libcvc4, so we declare them to be a private library. We don't want to include libutil in the linking process for libmain because the relevant parts are already linked into libcvc4 (calls to libutil functions in libcvc4 would be made to a different copy of the functions than calls to libutil functions in libmain, which is probably bad if they use local variables).
CVC4_PUBLIC makes a symbol visible after linking it into a library, so this is used by StatisticsRegistry and friends to make the relevant libutil functions linkable in libmain. Headers in libutil that should be usable anywhere (e.g. `cardinality.h` and `abstract_value.h`) should not include `cvc4_private_library.h` or `cvc4_private.h` and declare all functions and classes as CVC4_PUBLIC.
One possible solution that I see is to declare the safe_print functions to be CVC4_PUBLIC. I am not quite sure what the other solution would be (except making them part of another library or adding CVC4_PUBLIC wrapper functions just for the ones used in libutil).